plugins/optimization-detective/detect.js

```diff
@@ -505,28 +537,67 @@ export default async function detect( {
 		);
 	} );
 
+	// Only proceed with submitting the URL Metric if viewport stayed the same size. Changing the viewport size (e.g. due
+	// to resizing a window or changing the orientation of a device) will result in unexpected metrics being collected.
+	if (
+		window.innerWidth !== urlMetric.viewport.width ||
+		window.innerHeight !== urlMetric.viewport.height
+	) {
+		if ( isDebug ) {
+			log(
+				'Aborting URL Metric collection due to viewport size change.'
+			);
+		}
+		return;
+	}
+
 	if ( extensions.size > 0 ) {
+		/** @type {Promise[]} */
+		const extensionFinalizePromises = [];
+
+		/** @type {string[]} */
+		const finalizingExtensionModuleUrls = [];
+
 		for ( const [
 			extensionModuleUrl,
 			extension,
 		] of extensions.entries() ) {
 			if ( extension.finalize instanceof Function ) {
 				try {
-					await extension.finalize( {
-						isDebug,
-						getRootData,
-						getElementData,
-						extendElementData,
-						extendRootData,
-					} );
+					extensionFinalizePromises.push(
+						extension.finalize( {
+							isDebug,
+							getRootData,
+							getElementData,
+							extendElementData,
+							extendRootData,
+						} )
```

Comment:
This needs to account for an extension (erroneously) not implementing `finalize` as an async function, in which case it will not return a promise. It should check for a `Promise` return value.